{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 这个文件是下载并导入的文件"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "import datetime\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "import sys\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GPL96 = GEOparse.get_GEO(geo='GPL96', destdir=\"./datasets\", silent=True)\n",
    "GPL570 = GEOparse.get_GEO(geo='GPL570', destdir=\"./datasets\", silent=True)\n",
    "GPL571 = GEOparse.get_GEO(geo='GPL571', destdir=\"./datasets\", silent=True)\n",
    "GPL2507 = GEOparse.get_GEO(geo='GPL2507', destdir=\"./datasets\", silent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GSE_IDs = [ \"68310\",\n",
    "            \"6269\",\n",
    "            \"20346\",\n",
    "            \"21802\",\n",
    "            \"27131\",\n",
    "            \"28750\",\n",
    "            \"40012\",\n",
    "            \"40396\",\n",
    "            \"42026\",\n",
    "            \"57065\",\n",
    "            \"60244\",\n",
    "            \"63990\",\n",
    "            # \"66099\", https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE66099 This dataset thus includes all unique patients from GSE4607, GSE8121, GSE9692, GSE13904, GSE26378, and GSE26440.\n",
    "            # \"68310\", #  可能太大电脑爆炸 1610 名健康成年人 共有 880 个阵列，对应 133 个个体\n",
    "            '111368',\n",
    "            ]\n",
    "none_key=[\"Uninfected type 2 diabetes mellitus\",\"non-infectious illness\",\"Vac\",\"Healt\",\"Blood_Control\",\"Control_Healthy\",\"PAXgene whole blood, healthy control\",\"Blood_HV\",\"whole blood-Healthy Control\",\"healthy subject\",\"Uninfected healthy\",\"Uninfected healthy\",\"HC\",\"Vac\",\"pathogen: None\",\"WB-control\", \"disease: Control\"]\n",
    "bacteria_key=[\"bacterial\",\"bacterial pneumonia\",\"Experiment_Post_Surgical\",\"Experiment_Sepsis\",\"PAXgene whole blood, bacterial pneumonia\", \"Blood_P\",\"whole blood-BACTERIA\",\"intensive-care unit patient\",\"Septicemic melioidosis\",\"Other sepsis\",\"PBMC_S.aureus_MRSA_INF\",\"disease: Sepsis\",\"disease: SepticShock\",\"PBMC_S.pneumoniae_INF\",\"PBMC_S.aureus_MSSA_INF\",\"bacterial pneumonia_day\",\"pathogen: E.coli\",\"pathogen: Bacteria\",\"pathogen: MRSA\",\"pathogen: Salmonella\",\"pathogen: MSSA\",\"WB-bact\",\"PBMC_E.coli\"]\n",
    "virus_key=[\"infection: influenza A virus\",\"infection: influenza B virus\",\"infection: human coronavirus NL63\",\"infection: human coronavirus HKU1\",\"infection: human rhinovirus\",\"infection: enterovirus\",\"infection: respiratory syncytial virus A\",\"Vir\",\"viral\",\"Severe Influenza\",\"Pande\",\"blood_day\",\"PAXgene whole blood, SIRS\",\"PAXgene whole blood, influenza A pneumonia\",\"whole blood-VIRUS\",\"H1N1\",\"H3N2\",\"PBMC_InfluenzaA_INF\",\"PBMC_InfluenzaB_INF\",\"Severe Influenza_day\",\"pathogen: Adenovirus\",\"disease: SIRS\",\"pathogen: HHV6\",\"pathogen: Enterovirus\",\"pathogen: Rhinovirus\",\"WB-H1N1\",\"WB-RSV\",\"Adenovirus\",\"pathogen: Adenovirus\"]\n",
    "both_key=[\"PAXgene whole blood, mixed bacterial and influenza A pneumonia\",\"whole blood-COINFECTION\"]\n",
    "other_key=[\"infection: our tests did not detect one of the viruses sought\",\"Sarcoid\",\"Pneumonia\",\"Cancer\"]\n",
    "infection_types = {\n",
    "    \"both\": both_key,\n",
    "    \"bacteria\": bacteria_key,\n",
    "    \"virus\": virus_key,\n",
    "    \"other\": other_key,\n",
    "    \"none\": none_key\n",
    "}\n",
    "infection_map = {\n",
    "    \"non-infectious illness\": \"None\",\n",
    "    \"None\": \"None\",\n",
    "    \"Control\": \"None\",\n",
    "    \"viral\": \"virus\",\n",
    "    \"Adenovirus\": \"virus\",\n",
    "    \"HHV6\": \"virus\",\n",
    "    \"Enterovirus\": \"virus\",\n",
    "    \"Rhinovirus\": \"virus\",\n",
    "    \"bacterial_status: No\": \"virus\",\n",
    "    \"bacterial_status: Excluded\": \"virus\",\n",
    "    \"influenza A virus\": \"virus\",\n",
    "    \"influenza A virus and human rhinovirus\": \"virus\",\n",
    "    \"influenza A virus and human coronavirus OC43\": \"virus\",\n",
    "    \"influenza A virus and respiratory syncytial virus B\": \"virus\",\n",
    "    \"bacterial\": \"bacteria\",\n",
    "    \"E.coli\": \"bacteria\",\n",
    "    \"Bacteria\": \"bacteria\",\n",
    "    \"MRSA\": \"bacteria\",\n",
    "    \"Salmonella\": \"bacteria\",\n",
    "    \"MSSA\": \"bacteria\",\n",
    "    \"bacterial_status: N/A\": \"other\",\n",
    "    \"our tests did not detect one of the viruses sought\": \"other\",\n",
    "    \"bacterial_status: Yes\": \"both\",\n",
    "}\n",
    "value_map = {\n",
    "    \"Log2 normalized signal\": True,\n",
    "    \"RMA-normalised signal intensity from Partek Genomics Suite\": True,\n",
    "    \"RMA signal intensity (log base 2)\": True,\n",
    "    \"RMA\": True,\n",
    "    \"log2 quantile normalized\": True,\n",
    "    \"RMA normalized and log2 transformed\": True,\n",
    "    \"RMA normalized and Log2 transformed\": True,\n",
    "    \"gcRMA normalized gene expression values\": True,\n",
    "    \"Raw signal intensities were normalized by background adjustment, variance stabilization transformation, and robust spline normalization in R package lumi\": True,\n",
    "    \"MAS5-calculated Signal intensity\": False,\n",
    "    \"Average normalized\": False,\n",
    "    \"Normalized value from BeadStudio software\": False,\n",
    "    \"cubic spline normalised intensity\": False,\n",
    "    \"GenomeStudio quantile normalised average signal intensity\": False,\n",
    "    \"quantile-normalized signal intensity\": False,\n",
    "    \"Illumina calculated signal intensity [AVG_Signal]\": False,\n",
    "}\n",
    "\n",
    "platforms = [\"GPL570\", \"GPL571\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSE68310\n",
      "GSE6269\n",
      "GSE20346\n",
      "GSE21802\n",
      "GSE27131\n",
      "GSE28750\n",
      "GSE40012\n",
      "GSE40396\n",
      "GSE42026\n",
      "GSE57065\n",
      "GSE60244\n",
      "GSE63990\n",
      "GSE111368\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSE</th>\n",
       "      <th>GSM</th>\n",
       "      <th>GPL</th>\n",
       "      <th>description</th>\n",
       "      <th>value</th>\n",
       "      <th>infection</th>\n",
       "      <th>type</th>\n",
       "      <th>title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE68310</td>\n",
       "      <td>GSM1667382</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Raw signal intensities were normalized by back...</td>\n",
       "      <td>0         174.014428\n",
       "1         178.135385\n",
       "2   ...</td>\n",
       "      <td>virus</td>\n",
       "      <td>influenza A virus</td>\n",
       "      <td>WholeBloodRNA_JR0021_Baseline</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSE68310</td>\n",
       "      <td>GSM1667383</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Raw signal intensities were normalized by back...</td>\n",
       "      <td>0         178.375804\n",
       "1         182.438011\n",
       "2   ...</td>\n",
       "      <td>virus</td>\n",
       "      <td>influenza A virus</td>\n",
       "      <td>WholeBloodRNA_JR0021_Day0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSE68310</td>\n",
       "      <td>GSM1667384</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Raw signal intensities were normalized by back...</td>\n",
       "      <td>0         177.493280\n",
       "1         180.617391\n",
       "2   ...</td>\n",
       "      <td>virus</td>\n",
       "      <td>influenza A virus</td>\n",
       "      <td>WholeBloodRNA_JR0021_Day2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSE68310</td>\n",
       "      <td>GSM1667385</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Raw signal intensities were normalized by back...</td>\n",
       "      <td>0         175.914112\n",
       "1         182.514468\n",
       "2   ...</td>\n",
       "      <td>virus</td>\n",
       "      <td>influenza A virus</td>\n",
       "      <td>WholeBloodRNA_JR0021_Day4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSE68310</td>\n",
       "      <td>GSM1667386</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Raw signal intensities were normalized by back...</td>\n",
       "      <td>0        180.519156\n",
       "1        181.402787\n",
       "2     ...</td>\n",
       "      <td>virus</td>\n",
       "      <td>influenza A virus</td>\n",
       "      <td>WholeBloodRNA_JR0021_Day6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2452</th>\n",
       "      <td>GSE111368</td>\n",
       "      <td>GSM3029687</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Log2 normalized signal</td>\n",
       "      <td>0          26.090496\n",
       "1          22.641748\n",
       "2   ...</td>\n",
       "      <td>both</td>\n",
       "      <td>bacterial_status: Yes</td>\n",
       "      <td>LON0177_H1N1_T1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2453</th>\n",
       "      <td>GSE111368</td>\n",
       "      <td>GSM3029688</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Log2 normalized signal</td>\n",
       "      <td>0          41.539173\n",
       "1          22.578267\n",
       "2   ...</td>\n",
       "      <td>both</td>\n",
       "      <td>bacterial_status: Yes</td>\n",
       "      <td>LON0146_H1N1_T3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2454</th>\n",
       "      <td>GSE111368</td>\n",
       "      <td>GSM3029689</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Log2 normalized signal</td>\n",
       "      <td>0          35.475081\n",
       "1          27.385960\n",
       "2   ...</td>\n",
       "      <td>virus</td>\n",
       "      <td>bacterial_status: No</td>\n",
       "      <td>LON0113_H1N1_T2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2455</th>\n",
       "      <td>GSE111368</td>\n",
       "      <td>GSM3029690</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Log2 normalized signal</td>\n",
       "      <td>0          33.681343\n",
       "1          26.777165\n",
       "2   ...</td>\n",
       "      <td>other</td>\n",
       "      <td>bacterial_status: N/A</td>\n",
       "      <td>2027_HC_HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2456</th>\n",
       "      <td>GSE111368</td>\n",
       "      <td>GSM3029691</td>\n",
       "      <td>GPL10558</td>\n",
       "      <td>Log2 normalized signal</td>\n",
       "      <td>0          26.427758\n",
       "1          26.596127\n",
       "2   ...</td>\n",
       "      <td>other</td>\n",
       "      <td>bacterial_status: N/A</td>\n",
       "      <td>HC0022_HC_HC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2457 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSE         GSM       GPL  \\\n",
       "0      GSE68310  GSM1667382  GPL10558   \n",
       "1      GSE68310  GSM1667383  GPL10558   \n",
       "2      GSE68310  GSM1667384  GPL10558   \n",
       "3      GSE68310  GSM1667385  GPL10558   \n",
       "4      GSE68310  GSM1667386  GPL10558   \n",
       "...         ...         ...       ...   \n",
       "2452  GSE111368  GSM3029687  GPL10558   \n",
       "2453  GSE111368  GSM3029688  GPL10558   \n",
       "2454  GSE111368  GSM3029689  GPL10558   \n",
       "2455  GSE111368  GSM3029690  GPL10558   \n",
       "2456  GSE111368  GSM3029691  GPL10558   \n",
       "\n",
       "                                            description  \\\n",
       "0     Raw signal intensities were normalized by back...   \n",
       "1     Raw signal intensities were normalized by back...   \n",
       "2     Raw signal intensities were normalized by back...   \n",
       "3     Raw signal intensities were normalized by back...   \n",
       "4     Raw signal intensities were normalized by back...   \n",
       "...                                                 ...   \n",
       "2452                             Log2 normalized signal   \n",
       "2453                             Log2 normalized signal   \n",
       "2454                             Log2 normalized signal   \n",
       "2455                             Log2 normalized signal   \n",
       "2456                             Log2 normalized signal   \n",
       "\n",
       "                                                  value infection  \\\n",
       "0     0         174.014428\n",
       "1         178.135385\n",
       "2   ...     virus   \n",
       "1     0         178.375804\n",
       "1         182.438011\n",
       "2   ...     virus   \n",
       "2     0         177.493280\n",
       "1         180.617391\n",
       "2   ...     virus   \n",
       "3     0         175.914112\n",
       "1         182.514468\n",
       "2   ...     virus   \n",
       "4     0        180.519156\n",
       "1        181.402787\n",
       "2     ...     virus   \n",
       "...                                                 ...       ...   \n",
       "2452  0          26.090496\n",
       "1          22.641748\n",
       "2   ...      both   \n",
       "2453  0          41.539173\n",
       "1          22.578267\n",
       "2   ...      both   \n",
       "2454  0          35.475081\n",
       "1          27.385960\n",
       "2   ...     virus   \n",
       "2455  0          33.681343\n",
       "1          26.777165\n",
       "2   ...     other   \n",
       "2456  0          26.427758\n",
       "1          26.596127\n",
       "2   ...     other   \n",
       "\n",
       "                       type                          title  \n",
       "0         influenza A virus  WholeBloodRNA_JR0021_Baseline  \n",
       "1         influenza A virus      WholeBloodRNA_JR0021_Day0  \n",
       "2         influenza A virus      WholeBloodRNA_JR0021_Day2  \n",
       "3         influenza A virus      WholeBloodRNA_JR0021_Day4  \n",
       "4         influenza A virus      WholeBloodRNA_JR0021_Day6  \n",
       "...                     ...                            ...  \n",
       "2452  bacterial_status: Yes                LON0177_H1N1_T1  \n",
       "2453  bacterial_status: Yes                LON0146_H1N1_T3  \n",
       "2454   bacterial_status: No                LON0113_H1N1_T2  \n",
       "2455  bacterial_status: N/A                     2027_HC_HC  \n",
       "2456  bacterial_status: N/A                   HC0022_HC_HC  \n",
       "\n",
       "[2457 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.DataFrame(\n",
    "    columns=['GSE','GSM','GPL','description','value','infection','type','title']\n",
    ")\n",
    "for gseid in GSE_IDs:\n",
    "        GEONAME=\"GSE\" + gseid\n",
    "        print(GEONAME)\n",
    "        gse = GEOparse.get_GEO(geo=GEONAME, destdir=\"./datasets\", silent=True)\n",
    "\n",
    "        n = 0\n",
    "        no_new_gpl = 1\n",
    "\n",
    "        for i in gse.gsms:\n",
    "            n += 1\n",
    "\n",
    "            title = gse.gsms[i].get_metadata_attribute('title')\n",
    "\n",
    "            # 判断属于哪一类\n",
    "            metadata = gse.gsms[i].metadata\n",
    "            found = False\n",
    "            \n",
    "            characteristics_ch1 = gse.gsms[i].metadata['characteristics_ch1'] \n",
    "            useful_characteristics_ch1 = ['pathogen','infection_status','disease','infection']\n",
    "            status = []\n",
    "            for characteristics in characteristics_ch1:\n",
    "                label = characteristics.split(':')[0].strip()\n",
    "                if label in useful_characteristics_ch1:\n",
    "                    status.append(characteristics.split(':')[1].strip())\n",
    "                if label == 'bacterial_status':\n",
    "                    status.append(characteristics)\n",
    "            if len(status) == 1:\n",
    "                datatype = status[0]\n",
    "                if status[0] in [\"non-infectious illness\",\"None\",\"Control\"]:\n",
    "                    infection = \"None\"\n",
    "                elif status[0] in [\"viral\",\"Adenovirus\",\"HHV6\",\"Enterovirus\",\"Rhinovirus\",\"bacterial_status: No\",\"bacterial_status: Excluded\"] or (\"bacterial\" not in status[0] and (\"virus\" in  status[0] or \"influenza\" in  status[0])):\n",
    "                    infection = \"virus\"\n",
    "                elif status[0] in [\"bacterial\",\"E.coli\",\"Bacteria\",\"MRSA\",\"Salmonella\",\"MSSA\"]:\n",
    "                    infection = \"bacteria\"\n",
    "                elif status[0] in [\"bacterial_status: N/A\"]:\n",
    "                    infection = \"other\"\n",
    "                elif status[0] in [\"bacterial_status: Yes\"]:\n",
    "                    infection = \"both\"\n",
    "                else:\n",
    "                    print(status[0])\n",
    "            elif len(status) > 1:\n",
    "                print(status)\n",
    "            else:\n",
    "                infection = \"unknown\"\n",
    "                matched_word = \"\"\n",
    "                metadata_values = metadata.values()  # store metadata values in a separate variable\n",
    "                n_k = 0\n",
    "                for infection_type, keys in infection_types.items():\n",
    "                    for value in metadata_values:\n",
    "                        for k in keys:\n",
    "                            if k in value[0]:\n",
    "                                infection = infection_type\n",
    "                                matched_word = k\n",
    "                                break\n",
    "                        if infection != \"unknown\":\n",
    "                            break  # exit the loop once we find the infection type\n",
    "                    if infection != \"unknown\":\n",
    "                        break  # exit the loop once we find the infection type\n",
    "                datatype = matched_word\n",
    "                if infection == \"unknown\":\n",
    "                    print('unknown')\n",
    "                    print(metadata)\n",
    "\n",
    "            # 数据处理\n",
    "            value = gse.gsms[i].table[\"VALUE\"]\n",
    "            sample_gpl = gse.gsms[i].metadata['platform_id']\n",
    "            description = gse.gsms[i].columns[\"description\"][1]\n",
    "            \n",
    "            if description in value_map:\n",
    "                if value_map[description]:\n",
    "                    value = 2**value\n",
    "            elif sample_gpl in platforms:\n",
    "                value = 2**value\n",
    "            else:\n",
    "                print(gse.gsms[i].columns[\"description\"][1])\n",
    "                print(sample_gpl)\n",
    "                print('Now just use the raw value')\n",
    "                print('Samples: ', value.head())\n",
    "                print('-'*50)\n",
    "\n",
    "            data = data.append({\n",
    "                'GSE' : GEONAME,\n",
    "                'GSM' : metadata['geo_accession'][0],\n",
    "                'GPL' : sample_gpl[0],\n",
    "                'description' : gse.gsms[i].columns[\"description\"][1],\n",
    "                'value' : value,\n",
    "                'infection' : infection,\n",
    "                'type': datatype,\n",
    "                'title' : title},ignore_index=True)\n",
    "data\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "databygpl = data.groupby(['GPL'])\n",
    "for i in databygpl.groups.keys():\n",
    "    values = pd.DataFrame()\n",
    "    gpl = GEOparse.get_GEO(geo=i, destdir=\"./datasets\", silent=True)\n",
    "    save_data = pd.DataFrame()\n",
    "    metadata = pd.DataFrame()\n",
    "    for j in databygpl.groups[i]:\n",
    "        value = data.loc[j, 'value']\n",
    "        values = values.append(value, ignore_index=True)\n",
    "        # save data will save the ['GSE','GSM','GPL','description','infection','type','title'] and value we just get\n",
    "        metadata = metadata.append(data.loc[j, ['GSE','GSM','GPL','description','infection','type','title']], ignore_index=True)\n",
    "    save_data = pd.concat([metadata, values], axis=1)\n",
    "    # rename the columns to GB_ACC. The column name now is the order of the sample, which is not good for us. \n",
    "    try:\n",
    "        GB_ACC = gpl.table['GB_ACC']\n",
    "    except:\n",
    "        print('error in rename columns for gpl: ', i, ' with error: ', sys.exc_info()[0])\n",
    "        GB_ACC = gpl.table['ID']\n",
    "    # create a dict to rename the columns. If the GB_ACC is null or empty, then use the original column name.\n",
    "    save_data.rename(columns=dict(zip(save_data.columns[7:], GB_ACC)), inplace=True)\n",
    "\n",
    "    # 其中id与GB_ACC对应关系存在 gpl.table 这个表格中， id的列为ID， GB_ACC的列为GB_ACC。部分id对应的GB_ACC为空，就保留原来的id列名。\n",
    "    # try:\n",
    "    #     cols_dict = dict(zip(gpl.table['ID'], gpl.table['GB_ACC']))\n",
    "    #     save_data.rename(columns=cols_dict, inplace=True)\n",
    "    # except KeyError:\n",
    "    #     pass\n",
    "    # except:\n",
    "    #     # log the error\n",
    "    #     print('error in rename columns for gpl: ', i, ' with error: ', sys.exc_info()[0])\n",
    "    \n",
    "    # save_data = pd.concat([data.loc[databygpl.groups[i], ['GSE','GSM','GPL','description','infection','type','title']], values], axis=1)\n",
    "    save_data.to_csv('./datasets/' + i + '_save.csv', index=False)\n",
    "    # save to pickle\n",
    "    save_data.to_pickle('./datasets/' + i + '_save.pkl')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "igem",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
